src/accountsdb/manager.zig

```diff
@@ -20,260 +20,201 @@ const AccountFile = sig.accounts_db.accounts_file.AccountFile;
-                    // NOTE: this might be too much in production, not sure
-                    try self.flush_slots.append(allocator, unrooted_slot.*);
-                }
-            }
-        }
+        break :new_id try flushSlot(db, newly_rooted_slot);
+    };
 
-        const must_flush_slots = self.flush_slots.items.len > 0;
-        defer if (must_flush_slots) {
-            self.flush_slots.clearRetainingCapacity();
-            self.unclean_account_files.clearRetainingCapacity();
-            self.shrink_account_files.clearRetainingCapacity();
-        };
+    // TODO: audit this and related fields - not sure why this is an atomic to begin with
+    db.largest_flushed_slot.store(newly_rooted_slot, .monotonic);
+
+    // when we last made a full snapshot, and when we last made an incremental snapshot (relative to
+    // the last full snapshot).
+    const latest_full_snapshot_slot, const latest_inc_snapshot = blk: {
+        const snapshot_gen_info = db.latest_snapshot_gen_info.readCopy() orelse
+            break :blk .{ 0, 0 };
+        const incremental_info = snapshot_gen_info.inc orelse
+            break :blk .{ snapshot_gen_info.full.slot, 0 };
+        break :blk .{ snapshot_gen_info.full.slot, incremental_info.slot };
+    };
 
-        if (must_flush_slots) {
-            self.db.logger.debug()
-                .logf("flushing slots: min: {}...{}", std.mem.minMax(Slot, self.flush_slots.items));
+    // TODO: this should be configurable - leaving disabled until snapshot generation is fixed
+    const snapshot_generation_enabled = builtin.is_test;
 
-            // flush the slots
-            try self.unclean_account_files.ensureTotalCapacityPrecise(
-                allocator,
-                self.flush_slots.items.len,
-            );
+    const make_full_snapshot = snapshot_generation_enabled and newly_rooted_slot >=
+        latest_full_snapshot_slot + config.slots_per_full_snapshot;
+    const make_inc_snapshot = snapshot_generation_enabled and !make_full_snapshot and
+        newly_rooted_slot >=
+            latest_inc_snapshot + latest_full_snapshot_slot + config.slots_per_incremental_snapshot;
 
-            var largest_flushed_slot: Slot = 0;
-            for (self.flush_slots.items) |flush_slot| {
-                const unclean_file_id = flushSlot(self.db, flush_slot) catch |err| {
-                    // flush fail = loss of account data on slot -- should never happen
-                    self.db.logger.err()
-                        .logf("flushing slot {d} error: {s}", .{ flush_slot, @errorName(err) });
-                    continue;
-                };
-                self.unclean_account_files.appendAssumeCapacity(unclean_file_id);
-                largest_flushed_slot = @max(largest_flushed_slot, flush_slot);
-            }
-            _ = self.db.largest_flushed_slot.fetchMax(largest_flushed_slot, .seq_cst);
-        }
+    // TODO: might be a good idea to move snapshot creation to another thread
```

Comment:
Yeah, otherwise you're blocking replay. This seems pretty important actually.

Reply:
Yeah definitely. How about we put it onto a thread once we enable it properly?